### DISCUSSION TITLE
test the minio Java API's upload speed for 100M files in concurrent situations

### DISCUSSION THREAD

**Discussion by jiajinhui123** [minio/minio#13925] [Q&A] [answered]
Hello, the question I raised is not a bug, but a question I have
## Expected Behavior
the expectation is within 3S

## Current Behavior
Open multithreaded upload file 
1. the following is the code I call Java API
```
    Runnable runnable = new Runnable() {
          @Override
            public void run() {
                MinioClient minioClient = MinioClient.builder().endpoint("http://192.XXX.XXX.XXX:9001")
                        .credentials("minioadmin", "minioadmin").build();
                try {
                    Instant start = Instant.now();
                    FileInputStream fis = new FileInputStream("C:\\config\\resource\\100M\\elasticsearch.tar.gz");
                    Instant end1 = Instant.now();
                    System.out.println(Thread.currentThread().getName()+", 本地读文件用了："+ Duration.between(start,end1).toMillis());
                    minioClient.putObject(PutObjectArgs.builder()
                            .bucket("henghe-resource-test")
                            .object("es-100M-"+ FileUtils.getUUID() +".tar.gz")
                            .stream(new BufferedInputStream(fis),fis.available(),-1)
                            .build());
                    Instant end = Instant.now();
                    System.out.println(Thread.currentThread().getName()+", minio客户端上传 用了："+ 
                   Duration.between(end1,end).toMillis());

                    System.out.println(Thread.currentThread().getName()+", 总时间 用了："+ Duration.between(end1,end).toMillis());
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
    }
   Thread thread = null;
   for (int i = 0; i <50 ; i++) {
       thread = new Thread(runnable);
       thread.start();
    }
```
2. result 
```
Thread-20, minio客户端上传 用了：49381
Thread-35, minio客户端上传 用了：49901
Thread-6, minio客户端上传 用了：50210
Thread-5, minio客户端上传 用了：50526
Thread-14, minio客户端上传 用了：53389
Thread-40, minio客户端上传 用了：53519
Thread-17, minio客户端上传 用了：53554
Thread-23, minio客户端上传 用了：56910
Thread-21, minio客户端上传 用了：56943
Thread-43, minio客户端上传 用了：56964
Thread-30, minio客户端上传 用了：57673
Thread-3, minio客户端上传 用了：58037
Thread-48, minio客户端上传 用了：58127
Thread-25, minio客户端上传 用了：60182
Thread-46, minio客户端上传 用了：61271
Thread-10, minio客户端上传 用了：62884
Thread-19, minio客户端上传 用了：64128
Thread-18, minio客户端上传 用了：65553
Thread-44, minio客户端上传 用了：65624
Thread-2, minio客户端上传 用了：66697
Thread-13, minio客户端上传 用了：67975
Thread-38, minio客户端上传 用了：68106
Thread-7, minio客户端上传 用了：69032
Thread-27, minio客户端上传 用了：69085
Thread-31, minio客户端上传 用了：69962
Thread-24, minio客户端上传 用了：69980
Thread-36, minio客户端上传 用了：70012
Thread-1, minio客户端上传 用了：69981
Thread-8, minio客户端上传 用了：70021
Thread-26, minio客户端上传 用了：705144
Thread-9, minio客户端上传 用了：70764
Thread-49, minio客户端上传 用了：71481
Thread-0, minio客户端上传 用了：71516
Thread-39, minio客户端上传 用了：71533
Thread-41, minio客户端上传 用了：71697
Thread-33, minio客户端上传 用了：71896
Thread-16, minio客户端上传 用了：71896
Thread-47, minio客户端上传 用了：72446
Thread-29, minio客户端上传 用了：72463
Thread-42, minio客户端上传 用了：72495
Thread-12, minio客户端上传 用了：72491
Thread-32, minio客户端上传 用了：72536
Thread-4, minio客户端上传 用了：73051
Thread-22, minio客户端上传 用了：73053
Thread-37, minio客户端上传 用了：73052
Thread-11, minio客户端上传 用了：73074
Thread-34, minio客户端上传 用了：73076
Thread-45, minio客户端上传 用了：73053
Thread-28, minio客户端上传 用了：73063
Thread-15, minio客户端上传 用了：73063
```
3. I set the block size to -1, the default is 5M, I would like to ask, according to 5M blocks, is multithreaded upload, I did not find in the source
4. I deployed distributed Minio, and here is my deployment script
```
#!/bin/bash
export MINIO_ROOT_USER=minioadmin
export MINIO_ROOT_PASSWORD=minioadmin

/home/minio/run/minio server --address :9001 \
http://192.XXX.XXX.111/home/minio/data1 \
http://192.XXX.XXX.112/home/minio/data1 \
http://192.XXX.XXX.113/home/minio/data1 \
http://192.XXX.XXX.113/home/minio/data2 \
```
5.Server Information
![image](https://user-images.githubusercontent.com/83194758/146327419-72d2d4e1-3b07-4fe9-bfe4-5d6bd5596fa9.png)

![image](https://user-images.githubusercontent.com/83194758/146325831-ce812414-aab2-4db7-a2ce-44e766aa7728.png)

![image](https://user-images.githubusercontent.com/83194758/146325999-6425b232-d98f-48f2-87e0-9adbfd825fd8.png)

![image](https://user-images.githubusercontent.com/83194758/146326627-d489b5b0-f3b6-4923-a395-e743109f8324.png)

![image](https://user-images.githubusercontent.com/83194758/146326353-ef274605-1c0c-449f-97a0-0e2f2a46b588.png)

Please help me analyze why the upload speed is so slow in the case of concurrency. How should I adjust? Thank you very much

**Comment by klauspost** [comment, marked as answer]
I suspect your disks are the main bottleneck. You upload 50x100MB = 5GB. I never expect this to be done in 3 seconds with this small a setup.

You can use [warp](https://github.com/minio/warp) to see if you get similar numbers.

If you want us to go into more details, performance testing your software look at our [subscription offers](https://min.io/pricing) or write `hello@minio.io`.

**Reply by jiajinhui123** [reply]
Let me ask again. The default partition is 5M. Is the partition uploaded by multiple threads

**Reply by klauspost** [reply]
@jiajinhui123 I don't know the details of the java sdk. I don't think it matter much.

**Reply by balamurugana** [reply]
1. As you pass `-1` to optional parameter `partSize` in `.stream(new BufferedInputStream(fis),fis.available(),-1)`, `minio-java` computes suitable part size. The minimum part size is 5MiB. However you have an option to override default behavior.
2. `putObject()` method doesn't do parallel uploads. If you want parallel uploads of object parts, you could inherit `MinioClient` and upload data in parallel using lower level APIs.
